sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -96,43 +224,243 @@ public boolean isInputSortRelAndLimitOnly() {
 
   /** {@code CalcFn} is the executor for a {@link BeamCalcRel} step. */
   private static class CalcFn extends DoFn<Row, Row> {
-    private BeamSqlExpressionExecutor executor;
-    private Schema outputSchema;
+    private final String processElementBlock;
+    private final Schema outputSchema;
+    private transient @Nullable ScriptEvaluator se = null;
 
-    public CalcFn(BeamSqlExpressionExecutor executor, Schema outputSchema) {
-      super();
-      this.executor = executor;
+    public CalcFn(String processElementBlock, Schema outputSchema) {
+      this.processElementBlock = processElementBlock;
       this.outputSchema = outputSchema;
     }
 
+    ScriptEvaluator compile() {
+      ScriptEvaluator se = new ScriptEvaluator();
+      se.setParameters(
+          new String[] {outputSchemaParam.name, processContextParam.name, DataContext.ROOT.name},
+          new Class[] {
+            (Class) outputSchemaParam.getType(),
+            (Class) processContextParam.getType(),
+            (Class) DataContext.ROOT.getType()
+          });
+      try {
+        se.cook(processElementBlock);
+      } catch (CompileException e) {
+        throw new RuntimeException("Could not compile CalcFn: " + processElementBlock, e);
+      }
+      return se;
+    }
+
     @Setup
     public void setup() {
-      executor.prepare();
+      this.se = compile();
     }
 
     @ProcessElement
     public void processElement(ProcessContext c) {
-      Row inputRow = c.element();
-      @Nullable
-      List<Object> rawResultValues =
-          executor.execute(inputRow, null, BeamSqlExpressionEnvironments.forRow(inputRow, null));
-
-      if (rawResultValues != null) {
-        List<Object> castResultValues =
-            IntStream.range(0, outputSchema.getFieldCount())
-                .mapToObj(i -> castField(rawResultValues, i))
-                .collect(toList());
-        c.output(Row.withSchema(outputSchema).addValues(castResultValues).build());
+      assert se != null;
+      try {
+        se.evaluate(new Object[] {outputSchema, c, CONTEXT_INSTANCE});
+      } catch (InvocationTargetException e) {
+        throw new RuntimeException(
+            "CalcFn failed to evaluate: " + processElementBlock, e.getCause());
+      }
+    }
+  }
+
+  private static final Map<TypeName, Type> rawTypeMap =
+      ImmutableMap.<TypeName, Type>builder()
+          .put(TypeName.BYTE, Byte.class)
+          .put(TypeName.INT16, Short.class)
+          .put(TypeName.INT32, Integer.class)
+          .put(TypeName.INT64, Long.class)
+          .put(TypeName.FLOAT, Float.class)
+          .put(TypeName.DOUBLE, Double.class)
+          .build();
+
+  private Expression castOutput(Expression value, FieldType toType) {
+    if (value.getType() == Object.class) {
```

Comment:
Is this possible to get an expression with `Object` type? Would it be safer to throw here to avoid breaking something in our schema logic?

Reply:
See the comment below. This is the fast copy path, which is actually used quite a bit. When a field is copied from input to output without modification code is generated to fetch it with `getValue` which returns an object. That object can be passed through to the new row without any conversion.